TU20Bot/Configuration/AuthorizationModule.cs

```diff
@@ -0,0 +1,155 @@
+    }
+
+    public class AuthorizationModule : WebModuleBase {
+        private readonly Server server;
+        private readonly IWebModule module;
+        private const string bearerPrefix = "Bearer ";
+
+        private static void error(string text, IHttpResponse response) {
+            response.StatusCode = 400;
+            var writer = new StreamWriter(response.OutputStream);
+            writer.WriteLine(text);
+            writer.Flush();
+        }
+
+        protected override async Task OnRequestAsync(IHttpContext context) {
+            if (!ValidateToken(context.Request, context.Response)) {
+                throw HttpException.Unauthorized();
+            }
+
+            var authorization = context.Request.Headers["Authorization"];
+
+            // Extract the incoming JWT Token from the header
+            var token = authorization.Substring(bearerPrefix.Length);
+
+            Exception moduleError = null;
+
+            try {
+                var secret = Encoding.UTF8.GetBytes(server.config.jwtSecret);
+                var result = JWT.Decode(token, secret);
+
+                if (string.IsNullOrEmpty(result)) {
+                    error("ERROR: Empty payload.", context.Response);
+                    return;
+                }
+
+                // Deserialize the JWT Token into an AuthorizationPayload object
+                var payload = JsonConvert.DeserializeObject<AuthorizationPayload>(result);
+
+                if (payload == null) {
+                    error("ERROR: Invalid payload.", context.Response);
+                    return;
+                }
+
+                // If the token is expired
+                if (DateTime.Now.CompareTo(payload.validUntil) > 0) {
+                    error("ERROR: Expired token.", context.Response);
+                    return;
+                }
+
+                Console.WriteLine(
+                    "[Authorization] Access for {0} to \"{1}\".",
+                    payload.fullName, context.Request.Url);
+
+                // Direct the module to handle the request
+                try {
+                    await module.HandleRequestAsync(context);
+                } catch (Exception e) {
+                    moduleError = e;
+                }
+
+            } catch (IntegrityException) {
+                error("ERROR: Invalid JWT signature.", context.Response);
+            } catch (Exception) {
+                error("Error: Failure to parse JWT.", context.Response);
+            }
+
+            if (moduleError != null) {
+                throw moduleError;
+            }
+        }
+
+        /// <summary>
+        /// Check whether the incoming Authorisation header is not null and contains the expected bearer token
+        /// </summary>
+        /// <param name="request"></param>
+        /// <param name="response"></param>
+        /// <returns></returns>
+        public static bool ValidateToken(IHttpRequest request, IHttpResponse response) {
```

Comment:
Another comment about `ValidateToken` and `ValidatePermissions`. Instead of taking `IHttpRequest` and `IHttpResponse`, it might be better to go the more functional approach and instead take the authorization token and return the error message or null.

This increases the versatility of the methods (I can use them in places where I'm not necessarily processing an Http request).